openwisp_controller/config/tests/test_selenium.py

```diff
@@ -0,0 +1,253 @@
+                    (By.XPATH, f'//*[@value="{org2_default_template.id}"]')
+                )
+            )
+        except (TimeoutException, StaleElementReferenceException):
+            self.fail()
+
+        # org1 and shared templates should be visible
+        wait.until(
+            EC.visibility_of_any_elements_located(
+                (By.XPATH, f'//*[@value="{org1_required_template.id}"]')
+            )
+        )
+        wait.until(
+            EC.visibility_of_any_elements_located(
+                (By.XPATH, f'//*[@value="{org1_default_template.id}"]')
+            )
+        )
+        wait.until(
+            EC.visibility_of_any_elements_located(
+                (By.XPATH, f'//*[@value="{shared_required_template.id}"]')
+            )
+        )
+
+    def test_change_config_backend(self):
+        device = self._create_config(organization=self._get_org()).device
+        template = self._create_template()
+
+        self.login()
+        self.open(
+            reverse('admin:config_device_change', args=[device.id]) + '#config-group'
+        )
+        self.web_driver.find_element_by_xpath(f'//*[@value="{template.id}"]')
+        # Change config backed to
+        config_backend_select = Select(
+            self.web_driver.find_element_by_name('config-0-backend')
+        )
+        config_backend_select.select_by_visible_text('OpenWISP Firmware 1.x')
+        try:
+            WebDriverWait(self.web_driver, 1).until(
+                EC.invisibility_of_element_located(
+                    (By.XPATH, f'//*[@value="{template.id}"]')
+                )
+            )
+        except TimeoutException:
+            self.fail()
+
+    def test_template_context_variables(self):
+        self._create_template(
+            name='Template1', default_values={'vni': '1'}, required=True
+        )
+        self._create_template(
+            name='Template2', default_values={'vni': '2'}, required=True
+        )
+        device = self._create_config(organization=self._get_org()).device
+        self.login()
+        self.open(
+            reverse('admin:config_device_change', args=[device.id]) + '#config-group'
+        )
+        try:
+            WebDriverWait(self.web_driver, 2).until(
+                EC.text_to_be_present_in_element_value(
+                    (
+                        By.XPATH,
+                        '//*[@id="flat-json-config-0-context"]/div[2]/div/div/input[1]',
+                    ),
+                    'vni',
+                )
+            )
+        except TimeoutException:
+            self.fail()
+        self.web_driver.find_element_by_xpath(
+            '//*[@id="container"]/div[2]/a[3]'
+        ).click()
+        try:
+            WebDriverWait(self.web_driver, 2).until(EC.alert_is_present())
+        except TimeoutException:
+            pass
+        else:
+            self.web_driver.switch_to_alert().accept()
+            self.fail()
```

Comment:
Only this assertion fails without the JS fixes, did you know that?

Can you add a brief description for the failure please (eg: `self.fail('expected to see/find ...')`?
The same suggestion applies to all other occurrences of `self.fail()`.

Reply:
> Only this assertion fails without the JS fixes, did you know that?

Yes it fails without the fix!